Add explicit layerThicknessEdge variables

cbegeman opened the pull request:
This PR replaces the existing `layerThicknessEdge` variable with two variables: `layerThicknessEdgeMean` refers to the mean of the neighboring cell-centered `layerThickness`, and `layerThicknessEdgeFlux` refers to the `layerThicknessEdge` used to flux momentum and tracers. When `config_thickness_flux_type = 'centered'` `layerThicknessEdgeFlux = layerThicknessEdgeMean`, but when `config_thickness_flux_type = 'upwind'` `layerThicknessEdgeFlux` is upwinded. 

Prior to this PR, when `config_thickness_flux_type = 'upwind'` `layerThicknessEdge` was upwinded everywhere. This PR uses `layerThicknessEdgeFlux` for horizontal advection terms and `layerThicknessEdgeMean` elsewhere, such as for mixing parameterizations. This change is in preparation for adding additional options for thickness fluxes. 

This PR also removes `layerThicknessEdge` from global stats, since it did not offer much information not given by the global stats for `layerThickness`.

cbegeman commented on components/mpas-ocean/src/shared/mpas_ocn_vmix.F:
```diff
@@ -288,13 +287,10 @@ subroutine ocn_vel_vmix_tend_implicit_rayleigh(dt, kineticEnergyCell, vertViscTo
         N = maxLevelEdgeTop(iEdge)
         if (N .gt. 0) then
 
-         ! layerThicknessEdge is computed in compute_solve_diag, and is not available yet,
-         ! so recompute layerThicknessEdge here.
```
Noting that this statement is false. `layerThicknessEdge` is computed in `ocn_diagnostic_solve,` which is called for `timeLevel=2` in each time integration scheme before `ocn_vel_vmix_tend` is called.

cbegeman commented on components/mpas-ocean/src/shared/mpas_ocn_tracer_hmix.F:
```diff
@@ -95,7 +95,7 @@ subroutine ocn_tracer_hmix_tend(meshPool, layerThicknessEdge, layerThickness, zM
 
       real (kind=RKIND), dimension(:,:), intent(in) :: &
-         layerThicknessEdge, &!< Input: thickness at edge
+         layerThicknessEdge, &!< Input: flux-related thickness at edge
```
Note that there are some instances (such as this one) in which I merely noted which `layerThicknessEdge` should be supplied as inputs to routines, and provided that input when the routine is called, rather than changing the variable name to `layerThicknessEdgeFlux` or `layerThicknessEdgeCenter` everywhere. If preferred, I could do the latter instead.

xylar replied:
I think it would make searching the code easier if these were all changed to `layerThicknessEdgeFlux` or `layerThicknessEdgeCenter`(`ed`).  But it's not a strong opinion.

cbegeman commented on components/mpas-ocean/src/shared/mpas_ocn_gm.F:
```diff
@@ -643,7 +643,7 @@ subroutine ocn_GM_compute_Bolus_velocity(statePool, &
                  BruntVaisalaFreqTopEdge = 0.5_RKIND*(max(BruntVaisalaFreqTop(k,cell1),0.0_RKIND) + &
                                           max(BruntVaisalaFreqTop(k,cell2),0.0_RKIND))
-                 shearEdgeInv = (0.5_RKIND*(layerThickEdge(k-1,iEdge) + layerThickEdge(k,iEdge)))**2.0 &
+                 shearEdgeInv = (0.5_RKIND*(layerThickEdgeFlux(k-1,iEdge) + layerThickEdgeFlux(k,iEdge)))**2.0 &
```
@vanroekel Suppose we are employing the upwinded `layerThicknessEdge` for momentum and tracer fluxes (here designated `layerThickEdgeFlux`). Can you weigh in on whether it is appropriate to use the central-differenced `layerThicknessEdge` (here designated `layerThickEdgeCenter`) for the GM bolus velocity computation? Here, I only retained `layerThickEdgeFlux` in the computation of the Brunt-Vaisala frequency.

vanroekel replied:
@cbegeman I think this one has to be center averaged since it is for thickness weighted averaging on edges.

I think the `BruntVaisalaFreqTopEdge` needs to be center averaged as well.  It doesn't make sense to me to use the upwind layerThickness there either.  But perhaps I'm not fully understanding your last sentence of the comment.

cbegeman replied:
@vanroekel Thanks for the prompt feedback! I'm glad we're in agreement on center averaging in general for GM.

I was thinking that perhaps we'd want to be using the upwinded `layerThicknessEdge` to compute the vertical shear when the volume fluxes through the edge are `u*h_upwind`. If that doesn't make sense, I'm happy to make the change to centered thicknesses throughout this routine.

vanroekel replied:
I see, thanks for the explanation @cbegeman.  Given this is to compute the Richardson number I think we should be consistent with the BVF edge and Shear.  Although it may be interesting to explore different options and associated sensitivities.  I definitely see your point though, but I'd like to leave centered for now until we can do testing of how sensitive the calculation is to that choice in isolation.

cbegeman replied:
@vanroekel That sounds reasonable to me. We'll go with centered thickness throughout.

xylar commented:
@cbegeman, and others, a general question for discussion. Are we okay with `layerThicknessEdgeCenter` as opposed to `layerThicknessEdgeCentered`?  My first instinct was that it seemed like a layer thickness at the center of an edge as opposed to one found by centered averaging (I don't think it's actually correct that this layer thickness is from centered differences, since it's a mean rather than a difference).

cbegeman commented:
@xylar I can change it to `layerThicknessEdgeCentered` if you think that's more clear. To me, `layerThicknessEdgeCentered` could also refer to the layer thickness at the center of the edge. Are you suggesting that `layerThicknessEdgeMean` might be a better name?

xylar commented on components/mpas-ocean/src/shared/mpas_ocn_tendency.F:
```diff
@@ -424,7 +424,7 @@ subroutine ocn_tend_vel(tendPool, statePool, forcingPool, &
       call ocn_vel_forcing_tend(normalVelocity, sfcFlxAttCoeff, &
                                 sfcStress, kineticEnergyCell, &
-                                layerThickEdge, tendVel, err)
+                                layerThickEdgeCenter, tendVel, err)
```
My recollection from our conversation yesterday or last week was that we would likely want the upwind version in bottom drag because we would want bottom drag to be one of the terms that remains non-zero as we approach the thin film.  But I will defer to the folks you actually asked ;-)

xylar commented:
Yes, I think I like `layerThicknessEdgeMean`.   That's a great suggestion.  Let's see if others have feedback before you change anything.

vanroekel commented:
I very much like `layerThicknessEdgeMean`  I'd vote for that instead of Center or Centered.

dengwirda commented:
Just stepping back from the naming conversation for a moment, I'd say the main reason we want to support upwind-type `layerThicknessEdge` variables is to ensure we have monotonic thickness advection in certain cases --- in other words, that it shouldn't be possible numerically to create -ve thickness layers.

I think the main terms we have to worry about are the `div(u*h_flux)` and `div(u*h_flux*psi)` tendencies in the thickness and tracer advection steps, and perhaps also the PV-flux tendencies in the momentum eqn (though we should look into the PV case in more detail).

To aim for a minimal change, keeping all other forcing, friction and parameterisation tendencies using the existing centred scheme seems like a good place to start to me.

In terms of `layerThicknessEdgeFlux` itself, the intention is for this to be a blend of the centred and upwind thicknesses --- equivalent to the existing centred thickness in the deep ocean (so that we recover the energy conservation of TRiSK for global config.'s) but degrading to the upwind thickness in shallow regions to ensure we have stability.

cbegeman commented:
@dengwirda Thanks for posting this context. I had in mind that this PR would simply introduce the new variables and aim for correctness in designating which of these variables should be deployed in which terms.

To my knowledge folks aren't utilizing the `config_thickness_flux_type='upwind'` option, so all the tendencies will still be identical when `config_thickness_flux_type='centered'`. In light of this, it would be preferable to me to determine whether `layerThickEdgeFlux` or `layerThickEdgeCentered` is more appropriate in the forcing and friction terms and include that in this PR. If you think that decision should only be made after testing in coastal configurations, then I suppose we would have to postpone it to the next PR. What do you think?

dengwirda commented:
@cbegeman Yes, agree all round --- I feel that using `layerThicknessEdgeFlux` just in the `div(u*h)` and `div(u*h*psi)` tendencies to start with is a good option, with all other forcing, friction and parameterisation terms using the existing centred scheme. I think that's the most minimal change we can make that we can then extend for robust wetting + drying.

cbegeman commented:
@dengwirda Got it. Commits https://github.com/E3SM-Ocean-Discussion/E3SM/pull/12/commits/c731e62c96d85d6704ee8878efdb09fa8ec8d5a7 and https://github.com/E3SM-Ocean-Discussion/E3SM/pull/12/commits/454069fb891ce32df5866769ab25c64bb2126887 revert my changes to horizontal mixing and cvmix so they will use the centered thicknesses.

xylar commented on components/mpas-ocean/src/shared/mpas_ocn_diagnostics_variables.F:
```diff
@@ -672,7 +675,8 @@ subroutine ocn_diagnostics_variables_init(domain, jenkinsOn, hollandJenkinsOn, e
          !$acc                   GMStreamFuncZ,                   &
-         !$acc                   layerThickEdge,                  &
+         !$acc                   layerThickEdgeFlux,              &
+         !$acc                   layerThickEdgeMean,            &
```
```suggestion
         !$acc                   layerThickEdgeMean,              &
```

xylar commented on components/mpas-ocean/src/shared/mpas_ocn_diagnostics_variables.F:
```diff
@@ -909,7 +913,8 @@ subroutine ocn_diagnostics_variables_destroy(err) !{{{
          !$acc                   GMStreamFuncZ,                   &
-         !$acc                   layerThickEdge,                  &
+         !$acc                   layerThickEdgeFlux,              &
+         !$acc                   layerThickEdgeMean,            &
```
```suggestion
         !$acc                   layerThickEdgeMean,              &
```

xylar commented on components/mpas-ocean/src/shared/mpas_ocn_diagnostics_variables.F:
```diff
@@ -1107,7 +1112,8 @@ subroutine ocn_diagnostics_variables_destroy(err) !{{{
                  GMStreamFuncZ,                   &
-                 layerThickEdge,                  &
+                 layerThickEdgeFlux,              &
+                 layerThickEdgeMean,            &
```
```suggestion
                 layerThickEdgeMean,              &
```

xylar reviewed (changes requested):
It looks like there are still a few changes related to the move from `layerThciknessEdgeCenter` to `layerThicknessEdgeMean`.  I'll take a more in-depth look early next week.

Did the allocation issue get resolved or is that something I should help look into?

cbegeman commented:
@xylar Thanks for those suggestions! They're addressed in https://github.com/E3SM-Ocean-Discussion/E3SM/pull/12/commits/53e1542886b220f895e57f21ef7dec2a0027204e

> Did the allocation issue get resolved or is that something I should help look into?

I found the culprit typo. Thanks!

xylar commented on components/mpas-ocean/src/analysis_members/Registry_global_stats.xml:
```diff
@@ -71,7 +71,7 @@
 				 description="Minimum global value of tangentialVelocity on ocean edges."
 			/>
 			<var name="layerThicknessEdgeMin" array_group="mins" units="m"
-				 description="Minimum global value of layerThicknessEdge on ocean edges."
+				 description="Minimum global value of layerThicknessEdgeFlux on ocean edges."
```
I think this is a bit confusing.  I think we need probably either need `layerThicknessEdgeFluxMin`, etc. and `layerThicknessEdgeMeanMin` (also very confusing) or we need to just remove `layerTicknessEdge*` from the global stats.  @mark-petersen, do you have an opinion on this?

cbegeman replied:
I have added global stats for both `layerThicknessEdgeFlux` and `layerThicknessEdgeMean`.

xylar reviewed (changes requested):
A few more small things to change.  This is looking good!

cbegeman commented on components/mpas-ocean/src/shared/mpas_ocn_diagnostics.F:
```diff
@@ -610,8 +613,9 @@ subroutine ocn_diagnostic_solve_layerThicknessEdge(normalVelocity, layerThicknes
       !
       !-----------------------------------------------------------------
 
-      real (kind=RKIND), dimension(:,:), intent(out) :: &
-         layerThicknessEdge     !< Output: layer thickness on edge
+      real (kind=RKIND), dimension(:,:), intent(inout) :: &
+         layerThicknessEdgeMean, & !< Output: centered layer thickness on edge
+         layerThicknessEdgeFlux        !< Output: centered layer thickness on edge
```
@xylar You're right, they should be `intent(out)`. Thanks for catching that.

cbegeman commented on components/mpas-ocean/src/shared/mpas_ocn_vel_hadv_coriolis.F:
```diff
@@ -214,7 +214,7 @@ subroutine ocn_vel_hadv_coriolis_tend(normRelVortEdge, &
                qArr(k,iEdge) = qArr(k,iEdge) + &
                                edgeWeight*normalVelocity(k,eoe)* &
-                               avgVorticity*layerThickEdge(k,eoe)
+                               avgVorticity*layerThickEdgeFlux(k,eoe)
```
My understanding is that it should be the flux term here and that it won't conserve energy in the upwinded regions.

cbegeman commented:
@xylar Thanks so much for the detailed review! I believe I have made all the changes you suggested, including changing `layerThicknessEdge` everywhere to either `layerThickEdgeMean` or `layerThickEdgeFlux` so that these variables are easily searchable.

cbegeman commented on components/mpas-ocean/src/shared/mpas_ocn_tendency.F:
```diff
@@ -658,7 +658,7 @@ subroutine ocn_tend_tracer(tendPool, statePool, forcingPool, &
          normalThicknessFlux(k,iEdge) = &
                  normalTransportVelocity(k,iEdge)* &
-                 layerThickEdge(k, iEdge)
+                 layerThickEdgeFlux(k, iEdge)
```
@dengwirda Pointing you to the use of `layerThickEdgeFlux` in `ocn_tend_tracer` for your review.

cbegeman commented on components/mpas-ocean/src/shared/mpas_ocn_tendency.F:
```diff
@@ -1411,7 +1411,7 @@ subroutine ocn_tend_freq_filtered_thickness(tendPool, statePool, &
 
             do k = 1, maxLevelEdgeTop(iEdge)
-               flux = layerThickEdge(k,iEdge)*normalVelocity(k,iEdge)* &
+               flux = layerThickEdgeFlux(k,iEdge)*normalVelocity(k,iEdge)* &
```
@dengwirda Pointing you to the use of `layerThickEdgeFlux` in `ocn_tend_tracer` for your review.

xylar replied:
I don't think this one is in `ocn_tend_tracer`, it's in `ocn_tend_freq_filtered_thickness`.  But it still deserves a careful look.

cbegeman replied:
Good catch. @dengwirda Feel free to ignore :)

xylar reviewed (commented):
I will run the `pr` test suite tomorrow on a merge of this branch with `master`, using `master` as a baseline to make sure everything is bit-for-bit.  Assuming that test goes well, I think this PR is ready to move to E3SM-Project/E3SM.

cbegeman commented:
@xylar Thanks! In the meantime I can make those additions to global stats.

xylar commented:
@cbegeman, I ran into conflicts when I tried to merge this into `E3SM-Project/E3SM/master`.  I'll try to resolve them locally but I'd prefer if you do the rebase here when you get a chance.

xylar reviewed (approved):
This looks good to me!  A merge of this branch (fixing the conflict I already mentioned) passes the `pr` test suite suite on Anvil with Intel and Intel-MPI, using the current `E3SM-Project/E3SM/master` as a baseline.

cbegeman commented:
@xylar I rebased and added the additional global stats variables. The only testing I've done with this latest commit is the global analysis test to ensure that the new global stats are present.

xylar commented:
@cbegeman, I'll be happy to do additional testing (including global stats) once this is moved to E3SM-Project.  Thanks for the hard work!

cbegeman commented:
@xylar Thanks again for your review!

mark-petersen commented:
@cbegeman thanks for your detailed work here, and apologies for the delayed reponse. I agree with these changes. Having `layerThicknessEdgeMean` and `layerThicknessEdgeFlux` is very clear. 

The global stats names follow the previous patterns so I agree with your choices, even though `MeanMin` etc can be confusing. In general I think we should greatly reduce the number of variables in global stats (I only ever look at prognostic variables and KE), but that is a PR for another day. Do global statistics for `layerThicknessEdgeMean` and `layerThicknessEdgeFlux` provide more useful information than global stats for `layerThicknessEdge`?

I do worry about the memory footprint of MPAS-Ocean. We add 3D arrays freely, so the memory footprint becomes quite bloated. I'm guilty of that too. For example, it appears that `layerThicknessEdgeMean` is used in diffusive terms, forcing terms, and parameterizations.  Again this is not the purpose of this PR, but we could decide if the extra 3D array is really worth the savings of averaging  `layerThickness` at the neighboring cells every time we need it.

For the purpose of having a separate variable for the flux quantity for the layer thickness, defined with flags at the correct place, this all looks right to me.

cbegeman commented:
@mark-petersen Thanks for giving this a thorough look!

> Do global statistics for layerThicknessEdgeMean and layerThicknessEdgeFlux provide more useful information than global stats for layerThicknessEdge?

I personally don't see either of these variables being particularly useful from a global stats perspective because they are going to reflect the global stats for `layerThickness` very closely.

To your question about whether an additional 3d array is worth it: I guess the short answer is that we'd have to assess this further. If we're going to go this route, my first inclination would be to consider making `layerThicknessEdgeFlux` temporary (allocated, computed via diagnostic routine, and deallocated in time integration routines and ocn_tendency). I imagine this would be less costly than averaging in all the places that `layerThicknessEdgeMean` is needed, but we'd probably have to time it out to know for sure. 

Should we try to flag these issues for further consideration, with a github issue or somewhere in the code?

mark-petersen commented:
OK, since you agree, please take `layerThicknessEdgeMean` and `layerThicknessEdgeFlux` out of global stats now. That is relevant to this PR, and they don't add value to global stats. Sorry about your time spent putting them in.

The decision between permanently allocated arrays and flops is a tricky one. Flops are free, as they say, but retrieval of variables from memory takes much longer. Here we can either retrieve one column of `layerThicknessEdgeMean`; or retrieve two columns of `layerThickness` that are not contiguous in memory, and average them.  The first one might be faster because we retrieve half as many bytes from memory (@philipwjones any opinion on that?).

I think `layerThicknessEdgeFlux` is worth having in permanently allocated arrays, because it depends on the numerical method and needs to be consistent across the model.

So my conclusion is, go ahead and leave those two variables as-is. Thanks!

philipwjones commented:
You'll probably regret tagging me since this one hits two of my biggest pet peeves with MPAS and you're catching me after a long day of virtual workshop. So... <rant> (1) Registry must die and (2) if you've heard of software containers, ocean diagnostics is a software frickin' dumpster and adding any more to it is contributing to the dumpster fire that will result.  Can we please start properly scoping variables?  That's not directed at any of you specifically, just that we've collectively gotten lazy on this.</rant>

But in answer to the specific question, I'm not sure how the new edge thickness will be used so it's hard for me to say what's best. As a general rule, something as simple as this should be kept local and recomputed (even as a local scalar) rather than stored. But if it's used in more than a few places or needs to be consistent (like the edge flux) then it makes sense to compute once and share. Or if you're going to introduce options that require branching every time you use it, then computing once to avoid excessive branching might be appropriate. And if we're going to have both of these as shared variables, then maybe we need a thickness module so they can be explicitly use'd and dependencies are clear rather than continuing to add to the diagnostics clutter.

philipwjones commented:
Funny - I had a rant  /rant with brackets around much of that first paragraph but this GitHub editor interpreted it as formatting and left it off...

cbegeman commented:
@philipwjones @mark-petersen Thanks for this discussion. It seems to me that this change to where `layerThicknessEdge` variables are computed/stored is better suited to a separate pull request. The main goal of this PR was to remove ambiguity about where upwinded `layerThicknessEdge` should be used. I understand that this pull request does contribute to the `ocn_diagnostics_variables` dumpster, and I am willing to work on a separate PR to remove both `layerThicknessEdge` variables from `ocn_diagnostics_variables` if you feel that that's a priority. 

Regarding this next PR, I imagine that a separate thickness module would be appropriate so that we only compute `layerThicknessEdgeFlux` once, given that there is some branching here and given that the upwinded routine has if-clauses to determine the upwind direction. This next PR could also include moving the `layerThicknessFluxMean` computation locally.

mark-petersen reviewed (approved):
I agree with these changes. I think it is ready for an E3SM PR. Thanks @cbegeman!

cbegeman commented:
Moved to https://github.com/E3SM-Project/E3SM/pull/4832
